File: src/micapolos/tata8/model/Key.java
package micapolos.tata8.model;

import static micapolos.tata8.Game.keys;
import static micapolos.tata8.model.Span.span;

public enum Key implements Showable {
  LEFT(keys.left),
  RIGHT(keys.right),
  UP(keys.up),
  DOWN(keys.down),
  Z(keys.z),
  X(keys.x);

  private final micapolos.tata8.Key state;

  public final Event press;
  public final Event release;
  public final Boolean isPressed;

  Key(micapolos.tata8.Key state) {
    this.state = state;
    this.press = Event.with(state::pressed);
    this.release = Event.with(state::released);
    this.isPressed = Boolean.bool(state::isPressed);
  }

  public Span pressedSpan() {
    return span(isPressed, press, release);
  }

  @Override
  public String toString() {
    return String.format("Key.%s: %s", name(), state.isPressed ? "pressed" : "released");
  }

  static void main() {
    Key.LEFT.show();
  }
}

File: src/micapolos/ast/Key.kt
package micapolos.ast

import micapolos.tata8.Game

enum class Key(internal val tata8: micapolos.tata8.Key) {
  LEFT(Game.keys.left),
  RIGHT(Game.keys.right),
  UP(Game.keys.up),
  DOWN(Game.keys.down),
  Z(Game.keys.z),
  X(Game.keys.x);

  val isPressed: Expression<Boolean> =
    Expression.Application(
      Boolean::class,
      "Key.isPressed",
      listOf(constant(micapolos.ast.Key::class, this)))

  val pressed: Expression<Boolean> =
    Expression.Application(
      Boolean::class,
      "Key.pressed",
      listOf(constant(micapolos.ast.Key::class, this)))

  val released: Expression<Boolean> =
    Expression.Application(
      Boolean::class,
      "Key.released",
      listOf(constant(micapolos.ast.Key::class, this)))
}
